File: apps/agent/src/commands/execute.ts
import { spawn } from "node:child_process"

import type { AgentCommand, AgentCommandResult } from "@nms/shared"

/**
 * Platform actions used by allowlisted commands. Callers pass argv arrays
 * only — never a Hub-supplied string, and never `shell: true`.
 */
export type CommandRuntime = {
  spawnDetached: (file: string, args: readonly string[]) => Promise<void>
  platform: NodeJS.Platform
}

export function createCommandRuntime(
  platform: NodeJS.Platform = process.platform
): CommandRuntime {
  return {
    platform,
    spawnDetached(file, args) {
      return new Promise((resolve, reject) => {
        const child = spawn(file, [...args], {
          detached: true,
          stdio: "ignore",
          shell: false,
          windowsHide: true,
        })
        child.once("error", reject)
        child.once("spawn", () => {
          child.unref()
          resolve()
        })
      })
    },
  }
}

const LINUX_REBOOT = { file: "/sbin/shutdown", args: ["-r", "now"] } as const
const DARWIN_REBOOT = { file: "/sbin/shutdown", args: ["-r", "now"] } as const
const WINDOWS_REBOOT = {
  file: "shutdown.exe",
  args: ["/r", "/t", "0"],
} as const

const LINUX_RESTART = {
  file: "/bin/systemctl",
  args: ["restart", "lockhaven-agent.service"],
} as const
const DARWIN_RESTART = {
  file: "/bin/launchctl",
  args: ["kickstart", "-k", "system/com.lockhaven.agent"],
} as const
const WINDOWS_RESTART = {
  file: "sc.exe",
  args: ["stop", "LockhavenAgent"],
} as const

function rebootSpec(platform: NodeJS.Platform) {
  if (platform === "win32") return WINDOWS_REBOOT
  if (platform === "darwin") return DARWIN_REBOOT
  return LINUX_REBOOT
}

function restartSpec(platform: NodeJS.Platform) {
  if (platform === "win32") return WINDOWS_RESTART
  if (platform === "darwin") return DARWIN_RESTART
  return LINUX_RESTART
}

/**
 * Runs one allowlisted command. `command.kind` is the only input that
 * selects behavior; nothing from Hub is passed to a process.
 */
export async function executeAgentCommand(
  command: AgentCommand,
  runtime: CommandRuntime
): Promise<AgentCommandResult> {
  try {
    if (command.kind === "reboot") {
      const spec = rebootSpec(runtime.platform)
      await runtime.spawnDetached(spec.file, spec.args)
      return { id: command.id, status: "succeeded" }
    }

    if (command.kind === "restart") {
      const spec = restartSpec(runtime.platform)
      await runtime.spawnDetached(spec.file, spec.args)
      return { id: command.id, status: "succeeded" }
    }

    if (command.kind === "update") {
      return {
        id: command.id,
        status: "failed",
        detail: "No signed update is available.",
      }
    }

    return {
      id: command.id,
      status: "refused",
      detail: "This command is not allowed.",
    }
  } catch (error) {
    return {
      id: command.id,
      status: "failed",
      detail:
        error instanceof Error
          ? error.message.slice(0, 500)
          : "Command failed.",
    }
  }
}

export async function executeResolvedCommands(
  commands: AgentCommand[],
  runtime: CommandRuntime
) {
  const results: AgentCommandResult[] = []
  for (const command of commands) {
    results.push(await executeAgentCommand(command, runtime))
  }
  return results
}

File: apps/lockhaven-agent/internal/commands/execute.go
package commands

import (
	"os/exec"
	"runtime"

	"github.com/nic-southern/lockhaven/apps/lockhaven-agent/internal/service" // pragma: allowlist secret
)

type Command struct {
	ID   string
	Kind string
	Name string
}

type Result struct {
	ID     string `json:"id"`
	Status string `json:"status"`
	Detail string `json:"detail,omitempty"`
}

type Spec struct {
	File string
	Args []string
}

func RebootSpec(goos string) Spec {
	if goos == "" {
		goos = runtime.GOOS
	}
	if goos == "windows" {
		return Spec{File: "shutdown.exe", Args: []string{"/r", "/t", "0"}}
	}
	return Spec{File: "/sbin/shutdown", Args: []string{"-r", "now"}}
}

func RestartSpec(goos string) Spec {
	if goos == "" {
		goos = runtime.GOOS
	}
	if goos == "darwin" {
		return Spec{File: "/bin/launchctl", Args: []string{"kickstart", "-k", "system/com.lockhaven.agent"}}
	}
	if goos == "windows" {
		return Spec{
			File: "powershell.exe",
			Args: []string{
				"-NoProfile",
				"-NonInteractive",
				"-WindowStyle",
				"Hidden",
				"-Command",
				"Restart-Service -Name " + service.Name + " -Force",
			},
		}
	}
	return Spec{File: "/bin/systemctl", Args: []string{"restart", "lockhaven-agent.service"}}
}

type Runtime struct {
	GOOS           string
	SpawnDetached  func(file string, args []string) error
	Update         *UpdateOffer
	Services       []ServiceAllow
	RestartService func(target string) error
}

type ServiceAllow struct {
	Name   string
	Target string
}

func spawnDetached(file string, args []string) error {
	cmd := exec.Command(file, args...)
	cmd.Stdout = nil
	cmd.Stderr = nil
	cmd.Stdin = nil
	applySpawnAttr(cmd)
	return cmd.Start()
}

func DefaultRuntime() Runtime {
	return Runtime{
		GOOS:          runtime.GOOS,
		SpawnDetached: spawnDetached,
	}
}

func Execute(command Command, rt Runtime) Result {
	switch command.Kind {
	case "reboot":
		spec := RebootSpec(rt.GOOS)
		if err := rt.SpawnDetached(spec.File, spec.Args); err != nil {
			return Result{ID: command.ID, Status: "failed", Detail: err.Error()}
		}
		return Result{ID: command.ID, Status: "succeeded"}
	case "restart":
		spec := RestartSpec(rt.GOOS)
		if err := rt.SpawnDetached(spec.File, spec.Args); err != nil {
			return Result{ID: command.ID, Status: "failed", Detail: err.Error()}
		}
		return Result{ID: command.ID, Status: "succeeded"}
	case "update":
		return executeUpdate(command, rt)
	case "restart_service":
		return executeRestartService(command, rt)
	default:
		return Result{ID: command.ID, Status: "refused", Detail: "This command is not allowed."}
	}
}
